File: 프로그래머스/1/1845. 폰켓몬/폰켓몬.cpp
#include <vector>
#include <map>
#include <iostream>
using namespace std;

int solution(vector<int> nums)
{
    int answer = 0;
    int max = nums.size() / 2;

    map<int, int> m;

    for (int i{ 0 }; i < nums.size(); i++) {
        m.insert({ nums[i], 1 });
    }

    for (int i{ 0 }; i < m.size(); i++) {
        if (answer < max) {
            answer++;
        }
    }

    return answer;
}

File: 프로그래머스/1/1845. 폰켓몬/폰켓몬.java
import java.util.*;

class Solution {
    public int solution(int[] nums) {
        int answer = 0;
        int size = nums.length; // 폰켓못의 개수
        
        // 폰켓몬 담을 map 선언
        Map<Integer, Integer> map = new HashMap<>();
        
        for(int num : nums) {
            map.put(num, map.getOrDefault(num, 0)+1);
        }
        
        if(map.size() >= size/2) {
            answer = size/2;
        }
        else {
            answer = map.size();
        }
        
        return answer;
    }
}